File: src/shared/model-auth.ts
export type ModelAuth = {
  mode: 'bearer' | 'custom';
  headers: Record<string, string>;
  body: Record<string, unknown>;
};
export const defaultModelAuth = (): ModelAuth => ({ mode: 'bearer', headers: {}, body: {} });
const reserved = new Set([
  'model',
  'messages',
  'input',
  'tools',
  'tool_choice',
  'functions',
  'function_call',
  'stream',
  'response_format',
]);
export function modelAuth(value: unknown, reservedFields = reserved): ModelAuth {
  if (!value || typeof value !== 'object' || Array.isArray(value))
    throw new Error('Enter a valid authentication configuration.');
  const auth = value as ModelAuth;
  if (!['bearer', 'custom'].includes(auth.mode))
    throw new Error('Choose an authentication method.');
  const object = (item: unknown): item is Record<string, unknown> =>
    !!item && typeof item === 'object' && !Array.isArray(item);
  if (
    !object(auth.headers) ||
    !object(auth.body) ||
    Object.keys(auth.headers).length > 16 ||
    JSON.stringify(auth).length > 24000
  )
    throw new Error(
      'Use up to 16 headers and an authentication configuration under 24,000 characters.',
    );
  const headers: Record<string, string> = {};
  for (const [name, value] of Object.entries(auth.headers)) {
    const lower = name.toLowerCase();
    if (
      !/^[!#$%&'*+.^_`|~0-9a-z-]+$/i.test(name) ||
      [
        'host',
        'content-length',
        'transfer-encoding',
        'connection',
        'content-type',
      ].includes(lower) ||
      lower in headers ||
      typeof value !== 'string' ||
      /[\r\n\x00]/.test(value) ||
      value.length > 8192 ||
      !value.trim()
    )
      throw new Error('Use unique valid header names and non-empty values without line breaks.');
    headers[lower] = value.trim();
  }
  if (auth.mode === 'bearer' && headers.authorization)
    throw new Error('Choose Custom authentication to supply your own Authorization header.');
  function validate(item: unknown, depth = 0) {
    if (depth > 8) throw new Error('Authentication body is too deeply nested.');
    if (object(item))
      for (const [key, child] of Object.entries(item)) {
        if (['__proto__', 'constructor', 'prototype'].includes(key))
          throw new Error('Invalid authentication body field.');
        validate(child, depth + 1);
      }
    else if (Array.isArray(item)) item.forEach((child) => validate(child, depth + 1));
    else if (item !== null && !['string', 'number', 'boolean'].includes(typeof item))
      throw new Error('Authentication body must be JSON.');
  }
  if (Object.keys(auth.body).some((key) => reservedFields.has(key.toLowerCase())))
    throw new Error('Authentication body fields cannot replace protocol input or tool settings.');
  validate(auth.body);
  return { mode: auth.mode, headers, body: structuredClone(auth.body) };
}
export function modelHeaders(apiKey: string, auth: ModelAuth): Record<string, string> {
  return {
    ...(auth.mode === 'bearer' && apiKey ? { authorization: `Bearer ${apiKey}` } : {}),
    ...auth.headers,
  };
}

File: agent/lib/model_auth.py
"""Private connection authentication, applied consistently to every model call."""
import copy
import json
import re


def auth_config(value=None, reserved_fields=None):
    value = value or dict(mode='bearer', headers={}, body={})
    if not isinstance(value, dict) or value.get('mode') not in ('bearer', 'custom'):
        raise ValueError('Invalid authentication configuration')
    headers, body = value.get('headers', {}), value.get('body', {})
    if not isinstance(headers, dict) or not isinstance(body, dict) or len(headers) > 16 or len(json.dumps(value)) > 24000:
        raise ValueError('Invalid authentication configuration')
    normalized = {}
    for name, item in headers.items():
        lower = name.lower()
        if not re.fullmatch(r"[!#$%&'*+.^_`|~0-9a-z-]+", lower) or lower in ('host', 'content-length', 'transfer-encoding', 'connection', 'content-type') or lower in normalized or not isinstance(item, str) or not item.strip() or len(item) > 8192 or any(char in item for char in ('\r', '\n', '\0')):
            raise ValueError('Invalid authentication headers')
        normalized[lower] = item.strip()
    if value['mode'] == 'bearer' and 'authorization' in normalized:
        raise ValueError('Use Custom authentication for an Authorization header')
    if reserved_fields is None:
        reserved_fields = ('model', 'messages', 'input', 'tools', 'tool_choice', 'functions', 'function_call', 'stream', 'response_format')
    if any(key.lower() in reserved_fields for key in body):
        raise ValueError('Authentication cannot replace model input or tools')
    def check(item, depth=0):
        if depth > 8:
            raise ValueError('Authentication body is too deeply nested')
        if isinstance(item, dict):
            for key, child in item.items():
                if key in ('__proto__', 'constructor', 'prototype'):
                    raise ValueError('Invalid authentication field')
                check(child, depth + 1)
        elif isinstance(item, list):
            for child in item:
                check(child, depth + 1)
    check(body)
    return dict(mode=value['mode'], headers=normalized, body=copy.deepcopy(body))


def http_headers(key, auth=None):
    auth = auth_config(auth)
    return {**({'authorization': 'Bearer ' + key} if key and auth['mode'] == 'bearer' else {}), **auth['headers']}


def model_auth(key, auth=None):
    auth = auth_config(auth)
    headers = http_headers(key, auth)
    # LiteLLM needs a nonempty key to avoid inheriting ambient credentials.
    # OpenAI's explicit omit sentinel removes that generated header for custom auth.
    if auth['mode'] == 'custom':
        from openai import omit
        headers['Authorization'] = headers.pop('authorization', omit)
    elif 'authorization' in headers:
        headers['Authorization'] = headers.pop('authorization')
    result = dict(api_key=key if key and auth['mode'] == 'bearer' else 'dextana-keyless')
    if auth['mode'] == 'custom' or auth['headers']:
        result['extra_headers'] = headers
    if auth['body']:
        result['extra_body'] = auth['body']
    return result
